src/main/scala/com/fulcrumgenomics/umi/CallMolecularConsensusReads.scala

```diff
@@ -0,0 +1,122 @@
+import dagr.commons.io.Io
+import dagr.commons.util.LazyLogging
+import dagr.sopt._
+import dagr.sopt.cmdline.ValidationException
+import htsjdk.samtools._
+import htsjdk.samtools.util.CloserUtil
+
+import scala.collection.JavaConverters._
+
+
+@clp(description =
+  """
+    |Calls consensus sequences from reads with the same UMI identifier.
+    |
+    |This tool assumes that reads with the same identifier are grouped together (consecutive in the file).
+  """,
+  group = ClpGroups.Umi)
+class CallMolecularConsensusReads
+( @arg(flag="i", doc="The input SAM or BAM file.") val input: PathToBam,
+  @arg(flag="o", doc="Output SAM or BAM file to write consensus reads.") val output: PathToBam,
+  @arg(flag="r", doc="Output SAM or BAM file to write reads not used.") val rejects: PathToBam,
+  @arg(flag="t", doc="The SAM attribute with the unique UMI identifier.") val tag: String = DefaultTag,
+  @arg(flag="p", doc="The Prefix all consensus read names") val readNamePrefix: Option[String] = None,
+  @arg(flag="R", doc="The new read group ID for all the consensus reads.") val readGroupId: String = "A",
+  @arg(flag="1", doc="The Phred-scaled error rate for an error prior to the UMIs being integrated.") val errorRatePreUmi: LogDouble = DefaultErrorRatePreUmi,
+  @arg(flag="2", doc="The Phred-scaled error rate for an error post the UMIs have been integrated.") val errorRatePostUmi: LogDouble = DefaultErrorRatePostUmi,
+  @arg(flag="q", doc="Cap the maximum base quality in the input (after shifting).") val maxBaseQuality: LogDouble = DefaultMaxBaseQuality,
+  @arg(flag="s", doc="Subtract this base quality from the input base qualities (prior to capping).") val baseQualityShift: Double = DefaultBaseQualityShift,
+  @arg(flag="N", doc="Mask (make 'N') consensus bases with quality less than this threshold.") val minConsensusBaseQuality: LogDouble = DefaultMinConsensusBaseQuality,
+  @arg(flag="M", doc="The minimum number of reads to produce a consensus base.") val minReads: Int = DefaultMinReads,
+  @arg(flag="Q", doc="The minimum mean base quality across a consensus base to output.") val minMeanConsensusBaseQuality: LogDouble = DefaultMinMeanConsensusBaseQuality,
+  @arg(flag="P", doc="Require a consensus call for both ends of a pair if true.") val requireConsensusForBothPairs: Boolean = DefaultRequireConsensusForBothPairs
+) extends FgBioTool with LazyLogging {
+
+  Io.assertReadable(input)
+  Seq(output, rejects).foreach(Io.assertCanWriteFile(_))
+  if (tag.length != 2)         throw new ValidationException("attribute must be of length 2")
+  if (errorRatePreUmi < Zero)  throw new ValidationException("Error rate pre UMI must be >= 0")
+  if (errorRatePostUmi < Zero) throw new ValidationException("Error rate post UMI must be >= 0")
+
+  /** TODO
+    * - what about when a read has an indel causing the rest of the bases to mismatch?
+    * - what about if a read is soft-clipped and the rest of the bases mismatch (ex. 30M120S) [second pass]?
```

Comment:
I thought we would just ignore soft-clipping and process the entire read vs. the other reads?